src/autoscaler.py

```diff
@@ -1,54 +1,99 @@
 import logging
+from ops.pebble import ExecError
 from pathlib import Path
 from typing import Dict
 import yaml
 
 from errors import JujuEnvironmentError
 
 logger = logging.getLogger(__name__)
+CONTROLLER = "scaler-controller"
 
 
 @dataclass
 class AutoScaler:
     controller_data: Dict[str, str] = field(default_factory=dict)
+    model_data: Dict[str, str] = field(default_factory=dict)
     secrets: Dict[str, str] = field(default_factory=dict)
     command: str = ""
 
-    def _build_command(self, juju_config, charm):
-        model, scale = juju_config["default_model_uuid"], juju_config["scale"]
+    def _build_command(self, config, charm):
+        model, scale = config["default_model_uuid"], config["scale"]
         node_groups = scale.nodes(default_model=model.cfg)
         if not node_groups:
             logger.info("Missing juju-scale config")
             raise JujuEnvironmentError("Waiting for Juju Configuration")
 
-        provider_args = "--cloud-provider=juju"
-        nodes_args = " ".join([f"--nodes {node}" for node in node_groups])
-        namespace_args = f"--namespace {charm.model.name.strip()}"
+        namespace = f"--namespace {charm.model.name.strip()}"
+        provider = "--cloud-provider=juju"
+        nodes = " ".join([f"--nodes {node}" for node in node_groups])
 
-        self.command = f"{self.binary} {provider_args} {nodes_args} {namespace_args}"
+        extra_args = config["extra_args"].key_values
+        extra = ""
+        if extra_args:
+            extra = " " + " ".join(
+                sorted(f"--{key}='{value}'" for key, value in extra_args.items())
+            )
+
+        self.command = f"{self.binary} {namespace} {provider} {nodes}{extra}"
         return self
 
-    def apply_juju(self, juju_config, charm):
-        self._build_command(juju_config, charm)
+    def apply(self, config, charm):
+        self._build_command(config, charm)
+
         self.secrets = {
-            "user": juju_config["username"],
-            "password": juju_config["password"],
+            "user": config["username"],
+            "password": config["password"],
         }
+
         self.controller_data = {
-            "api-endpoints": juju_config["api_endpoints"].endpoints,
+            "api-endpoints": config["api_endpoints"].endpoints,
             "uuid": str(uuid.uuid4()),  # juju CLI requires a random unique uuid
-            "ca-cert": juju_config["ca_cert"].decoded,
+            "ca-cert": config["ca_cert"].decoded,
         }
+
+        default_model_uuid = config["default_model_uuid"].cfg
+        if default_model_uuid:
+            self.model_data = {
+                "models": {"admin/controller": {}, "admin/default": {"uuid": default_model_uuid}},
+                "current-model": "admin/default",
+            }
+        else:
+            self.model_data = {
+                "models": {"admin/controller": {}},
+                "current-model": "admin/controller",
+            }
```

Comment:
this helps set the default models and controllers for the juju cli client.  It can probably be removed when the juju cli is no longer used.